packages/vite/src/node/optimizer/registerMissing.ts

```diff
@@ -100,49 +148,70 @@ export function createMissingImporterRegisterFn(
     // dependencies will be asigned this promise from this point
     depOptimizationProcessing = newDepOptimizationProcessing()
 
-    let newData: DepOptimizationMetadata | null = null
-
     try {
-      const optimizeDeps = await createOptimizeDepsRun(
-        server.config,
-        true,
-        false,
-        metadata,
-        newDeps,
-        ssr
-      )
-
-      const processingResult = await optimizeDeps.run()
+      const processingResult = await runOptimizeDeps(config, newDeps)
+
+      const newData = processingResult.metadata
+
+      // After a re-optimization, if the internal bundled chunks change a full page reload
+      // is required. If the files are stable, we can avoid the reload that is expensive
+      // for large applications. Comparing their fileHash we can find out if it is safe to
+      // keep the current browser state.
+      const needsReload =
+        metadata.hash !== newData.hash ||
+        Object.keys(metadata.optimized).some((dep) => {
+          return (
+            metadata.optimized[dep]?.fileHash !==
+            newData.optimized[dep]?.fileHash
```

Comment:
What happens if both sides of this comparison are undefined due to the optional chaining?

Reply:
They should always be defined, as we are only adding new deps and never removing old ones. Updated in https://github.com/vitejs/vite/pull/7379/commits/8a54e91c14b70c2db86e18fbbc86af7e8b683413